src/BlissSymbol.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2023 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+
+import { html } from "htm/preact";
+import { getSvgMarkupString } from "./SvgUtils";
+
+export type BciAvId = number | (string|number)[];
+
+type BlissSymbolProps = {
+  bciAvId: BciAvId,
+  label: string
+}
+
+export function BlissSymbol (props: BlissSymbolProps) {
+  const { bciAvId, label } = props;
+  const svgMarkupString = getSvgMarkupString(bciAvId);
+
+  // The coercion to `any` and assignment to `raw` is _only_ for the unit
+  // tests to avoid the error:
+  // "Argument of type 'any[]' is not assignable to parameter of type
+  // 'TemplateStringsArray.  Property raw is mssing ..."
+  // see: https://stackoverflow.com/questions/50706337/importing-html-to-typescript-to-use-as-templatestringliteral#answer-51012181
+  // Also TypeScript-ESLint does not allow explicit `any` types; override that
+  // rule for this case.
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const templateStringArray = [`${svgMarkupString}`] as any;
+  templateStringArray.raw = [`${svgMarkupString}`];
+
+  return html`
```

Comment:
I'm wondering if the returned html in a structure below would make more sense semantically?

```
<svg id="{uuid}" />
<lable for="{uuid}">{label-text}</label>
```

What do you think?

Reply:
There is a bigger issue here @cindyli .

First a `<label>` element cannot used since a `<label>` must reference an `<input>` element.  If there is a need for an accessible name, then either built-in svg  elements, attributes, or text content can be used; or use `aria-label` or `aria-labelledby`. (A short summary of the possibilities is in the [SVG-AAM](https://www.w3.org/TR/svg-aam-1.0/#h-note-9)).

But, the prior question is whether an accessible name is needed.  It is not needed in the case of the `ActionBmwCodeCell` component.  That component is a `button` whose text content serves as the accessible name.   Any other markup within the `<button>`[ is considered "presentational"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#all_descendants_are_presentational) or of no semantic significance. What is needed in this case is to block the `<svg>`.  Back when only circles were rendered for the SVG graphics, I added an aria `role="presentation` to the `<svg>` element to make it presentational.  According to the MDN link above, that should not be necessary, but I'll confirm that.  However, if the `BlissSymbol` component automatically added, for example, an `aria-labelledby` to the `<svg>` element, that would negate any presentational role and violate the rule that buttons have no meaningful children.

Now, `BlissSymbol` components are not used only for `ActionBmwCodeCell` components, and the SVG may not be presentational in other contexts.  How about adding a boolean property to the `props` that indicates whether the SVG is presentational or not, such as `isPresentational`?  If the SVG is not presentational, then the <svg> markup could be: `<svg aria-labelledby="{uuid}" .../>`, and the `div` that  currently contains the text label could be `<div><span id="{uuid}">{label}</span></div>`.

Putting it all together and forgetting about the `TemplateStringsArray` coercion, it would like something like:
```
if (isPresentational) {
  return html`
    <svg role="none" />
    <div>label</div>
  `;
else {
  return html`
    <svg aria-labelledby="{uuid}" ... />
    <div><span id="{uuid}">label</span></div>
  `;
}
```

Reply:
Sure, we can leave it as is until the need for an accessible name arises.  It occurred to me that the `ContentBmwEncoding` and `ContentBmwMessage` areas might need the non-presentational form since the symbols displayed there are not buttons.  What are they?

Reply:
For symbols displayed in the encoding area, I think they are representing the meaning expressed by their labels. For example, if we have a setting in the future that users can choose to only display symbols without displaying their label, those symbols should be announced in the label texts when being read out. My feeling is symbols in the display areas are also meaningful and not just for presentation. But I'm not sure if my feeling is correct.

What do you think, @klown?

Reply:
From the point of view of a Bliss user, the symbols are never just presentation.  They are everything.  But, from the point of view of accessibility APIs and ATs, they are useless.  Screen readers, for example, cannot speak Bliss at this point in time**.  They can only speak the label text.

@cindyli wrote:
> For symbols displayed in the encoding area, I think they are representing the meaning expressed by their labels.

I think you are correct, so that label text needs to be available to ATs.  If the symbols in the display areas were buttons, then that text would be available.  But they are not buttons and the text has to be exposed in some fashion.  One way is what I outlined above for the non-presentational case.  The question still is, what is their aria role?  The best I can come up with for now is that they are images and I believe SVG elements are images by default (I'll check) .  To be sure, here is a small change to the output for non-presentational `BlissSymbol`:
```
<svg role="img" aria-labelledby="{uuid}" ...  />
<div><span id="{uuid}">label</span></div>
```

> ... if we have a setting in the future that users can choose to only display symbols without displaying their label, those symbols should be announced in the label texts when being read out. 

That's relatively easy by styling the `<span>label</span>` text with `display:none` or perhaps `visibility:hidden`.  Using `@aria-labelledby` will still provide ATs with the label text even if it is not visually rendered.

** - screen readers cannot handle Bliss today, but might in the future.  At one point, emojis were not handled well by screen readers, but that has changed.  Early on the fix was to use `@alt` text or `@arial-label` or `@aria-labelledby`. Now VoiceOver at least has built in speech for many emojis with no need for alt text.  When Bliss is merged into unicode, there will be a standard consistent encoding for each symbol that can be matched to speech output.

Reply:
Followup re:
>  The question still is, what is their aria role? The best I can come up with for now is that they are images and I believe SVG elements are images by default (I'll check) 

I found a better alternative [`graphics-symbol`](https://w3c.github.io/graphics-aria/#graphics-symbol), but I do not know how well it is supported.  The documentation suggests to set the `role="graphics-symbol img"`.  Browsers use the first role value that they understand so it would default to `img`.

(Aside:  the default role for `svg` elements is _not_ `img` but [`graphic-document`](https://www.w3.org/TR/svg-aam-1.0/#details-id-66).  These symbols are not documents).

Reply:
Your explanation helps me understand better. Your suggestion about adding `isPresentational` flag makes lots of sense. One question to clarify, when Bliss symbols are displayed in the encoding area:

1. If the label is displayed with a symbol, the symbol should be set to "presentation" only;
2. If the label is hidden and only the symbol is displayed, the symbol should be set to "non-presentation".

Is my understanding correct?

Reply:
After thinking it through, "presentation" vs. "non-presentation" is unrelated to showing or hiding the label.

One issue is the aria role that the `<svg>` element has in different html contexts.  If the `<svg>` element is inside a `<button>` element, it is presentational and has `role="none"`.  That is the case whether the label is visible or not.

If the `<svg>` is in the encoding area, then it is not within a `<button>` element but inside a plain `<div>`.  Its role should be `graphics-symbol img`.  That role is also irrespective of whether the label is visible or not.

Regarding labels, an effective way to handle their visibility _and_ their exposure to ATs is to use `aria-labelledy`.  That's because visibility does not affect `aria-labelledby`.  The accessible name is the same regardless of whether the label text is hidden or not.

Finally, I did a little more research on `<svg>` elements inside `<button>` elements.  There is an issue that `role=none` will not be enough if there is any text inside the `<svg>` element.  If there is text it will be available to ATs even if the main `<svg>` element is marked as presentational; and a screen reader will speak any text inside the `<svg>` tree in that case.  I don't think that the Bliss-SVG-Builder produces anything other than shapes but, to be safe, the recommendation is to add `aria-hidden="true"` to the `<svg>` element here.

Putting it all together, the `BlissSymbol` component should have this logic:

```
if (isPresentational) {
  return html`
    <svg aria-hidden="true" ... />
    <div id="${uuid}">label</div>
  `;
} else {
  return html`
    <svg role="graphic-symbol img" aria-labelledby="${uuid}" > ... />
    <div id="${uuid}">label</div>
  `;
}
```

The ActionBmwCodeCell should change to add the presentational indicator, and an `aria-labelledby`, like so:

```
return html`
  <button id="${props.id}" class="actionBmwCodeCell" style="${gridStyles}" aria-labelledBy="${props.id}">
    <${BlissSymbol} bciAvId=${bciAvId} label=${label} isPresentational="true" uuid="${props.id}" />
  </button>
`;
```

Showing/hiding the label should be handled by SCSS. I have not put a lot of thought into what the selector should be.  Maybe a `blissSymbolLabelText` class or something like that?

Reply:
Thanks for the research and thorough explanation, @klown. Very helpful for me.

I think we probably don't need to worry about the css class for hidding/showing labels in BlissSymbol component for now until the related feature implementation arises.